dubbo-common/src/main/java/org/apache/dubbo/common/URL.java

```diff
@@ -169,87 +169,127 @@ public URL(String protocol, String username, String password, String host, int p
-        if (i >= 0) {
-            path = url.substring(i + 1);
-            url = url.substring(0, i);
-        }
-        i = url.lastIndexOf("@");
-        if (i >= 0) {
-            username = url.substring(0, i);
-            int j = username.indexOf(":");
-            if (j >= 0) {
-                password = username.substring(j + 1);
-                username = username.substring(0, j);
-            }
-            url = url.substring(i + 1);
-        }
-        i = url.lastIndexOf(":");
-        if (i >= 0 && i < url.length() - 1) {
-            if (url.lastIndexOf("%") > i) {
-                // ipv6 address with scope id
-                // e.g. fe80:0:0:0:894:aeec:f37d:23e1%en0
-                // see https://howdoesinternetwork.com/2013/ipv6-zone-id
-                // ignore
-            } else {
-                port = Integer.parseInt(url.substring(i+1));
+        private void parseUsernameAndPassword(){
+            int i = url.indexOf("/");
+            if (i >= 0) {
+                path = url.substring(i + 1);
                 url = url.substring(0, i);
             }
+            i = url.lastIndexOf("@");
+            if (i >= 0) {
+                username = url.substring(0, i);
+                int j = username.indexOf(":");
+                if (j >= 0) {
+                    password = username.substring(j + 1);
+                    username = username.substring(0, j);
+                }
+                url = url.substring(i + 1);
+            }
+        }
+
+        private void parsePort(){
+            int i = url.lastIndexOf(":");
+            if (i >= 0 && i < url.length() - 1) {
+                if (url.lastIndexOf("%") > i) {
+                    // ipv6 address with scope id
+                    // e.g. fe80:0:0:0:894:aeec:f37d:23e1%en0
+                    // see https://howdoesinternetwork.com/2013/ipv6-zone-id
+                    // ignore
+                } else {
+                    port = Integer.parseInt(url.substring(i+1));
+                    url = url.substring(0, i);
+                }
+            }
+        }
+
+        private void parseHost(){
+            if (url.length() > 0) host = url;
+        }
+
+        private void parse(){
+            this.parseParameters();
+            this.parseProtocol();
+            this.parseUsernameAndPassword();
+            this.parsePort();
+            this.parseHost();
+        }
+
+        private URL build(){
+            return new URL(protocol, username, password, host, port, path, parameters);
         }
-        if (url.length() > 0) host = url;
-        return new URL(protocol, username, password, host, port, path, parameters);
     }
 
+
+
+    /**
+     * Parse url string
+     *
+     * @param url URL string
+     * @return URL instance
+     * @see URL
+     */
+    public static URL valueOf(String url) {
+        if (url == null || (url = url.trim()).length() == 0) {
+            throw new IllegalArgumentException("url == null");

```

Comment:
How about make the msg of exception "url is empty!"

Reply:
got it